micro-rdk/src/common/conn/viam.rs

```diff
@@ -734,35 +774,58 @@ where
         mut app_client: Option<AppClient>,
     ) -> Result<(), errors::ServerError> {
         let wait = Duration::from_secs(1); // should do exponential back off
+
         loop {
-            if let Some(app_client) = app_client {
-                let mut app_client_tasks: FuturesUnordered<AppClientTaskRunner> =
-                    FuturesUnordered::new();
+            if let Some(app_client) = &app_client {
+                log::info!("starting execution of app client tasks");
+                let mut app_client_tasks: FuturesUnordered<_> = FuturesUnordered::new();
                 for task in &self.app_client_tasks {
-                    app_client_tasks.push(AppClientTaskRunner {
-                        app_client: &app_client,
-                        invoker: task,
-                        state: TaskRunnerState::Run {
-                            task: task.invoke(&app_client),
-                        },
+                    app_client_tasks.push({
+                        async move {
+                            let mut duration = task.get_default_period();
+                            loop {
+                                let _ = Timer::after(duration).await;
+                                match task.invoke(app_client).await {
+                                    Ok(Some(next_duration)) => {
+                                        duration = next_duration;
+                                    }
+                                    Ok(None) => {}
+                                    e @ Err(_) => {
+                                        break e;
+                                    }
+                                }
+                            }
+                        }
```

Comment:
This (I think) replicates the functionality of the `TaskRunnerState` and `AppClientTaskRunner` machinery, but without the need for pin projection and unsafe code. I've verified that a failure in any one task causes the `FuturesUnordered::next` call to return an error, and that all the tasks stop/drop with the `FuturesUnordered`, so I think I'm getting the intended behavior here.

There is an extra scope in there (at one point I was initializing some variables before creating the future, but then they went away.